src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,4971 @@
+            bool minimalRepresentation =
+                ruleSet == AsnEncodingRules.DER || ruleSet == AsnEncodingRules.CER;
+
+            // The ITU-T specifications tecnically allow lengths up to ((2^128) - 1), but
+            // since Span's length is a signed Int32 we're limited to identifying memory
+            // that is within ((2^31) - 1) bytes of the tag start.
+            if (minimalRepresentation && lengthLength > sizeof(int))
+            {
+                bytesRead = 0;
+                return false;
+            }
+
+            uint parsedLength = 0;
+
+            for (int i = 0; i < lengthLength; i++)
+            {
+                byte current = source[bytesRead];
+                bytesRead++;
+
+                if (parsedLength == 0)
+                {
+                    if (minimalRepresentation && current == 0)
+                    {
+                        bytesRead = 0;
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+
+                    if (!minimalRepresentation && current != 0)
+                    {
+                        // Under BER rules we could have had padding zeros, so
+                        // once the first data bits come in check that we fit within
+                        // sizeof(int) due to Span bounds.
+
+                        if (lengthLength - i > sizeof(int))
+                        {
+                            bytesRead = 0;
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                        }
+                    }
+                }
+
+                parsedLength <<= 8;
+                parsedLength |= current;
+            }
+
+            // This value cannot be represented as a Span length.
+            if (parsedLength > int.MaxValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            if (minimalRepresentation && parsedLength < MultiByteMarker)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            length = (int)parsedLength;
+            return true;
+        }
+
+        internal (Asn1Tag, int?) ReadTagAndLength(out int bytesRead)
+        {
+            if (TryPeekTag(_data.Span, out Asn1Tag tag, out int tagBytesRead) &&
+                TryReadLength(_data.Slice(tagBytesRead).Span, _ruleSet, out int? length, out int lengthBytesRead))
+            {
+                int allBytesRead = tagBytesRead + lengthBytesRead;
+
+                if (tag.IsConstructed)
+                {
+                    // T-REC-X.690-201508 sec 9.1 (CER: Length forms) says constructed is always indefinite.
+                    if (_ruleSet == AsnEncodingRules.CER && length != null)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
```

Comment:
hm, maybe never mind my previous comment. IT looks like the `Cryptography_Der_Invalid_Encoding` message is used for all rulesets. Could use a name change in that case then.

Reply:
I'm not sure what our rules are for renaming resources.  The message just says "ASN.1"